File: java/app/src/test/java/kickstarter/TestHelpers.java
/**
 * Copyright (c) 2024 Jeffrey Jonathan Jennings
 * 
 * @author Jeffrey Jonathan Jennings (J3)
 * 
 * 
 */
package kickstarter;

import java.time.*;
import java.time.format.*;
import java.util.*;

import kickstarter.model.*;


public class TestHelpers {
    private static Random random = new Random(System.currentTimeMillis());

    public static String generateAirportCode() {
        String[] airports = new String[] {
                "ATL", "DFW", "DEN", "ORD", "LAX", "CLT", "MCO", "LAS", "PHX", "MIA",
                "SEA", "IAH", "JFK", "EWR", "FLL", "MSP", "SFO", "DTW", "BOS", "SLC",
                "PHL", "BWI", "TPA", "SAN", "LGA", "MDW", "BNA", "IAD", "DCA", "AUS"
        };

        return airports[random.nextInt(airports.length)];
    }

    public static String generateString(int size) {
        final String alphaString = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

        StringBuilder sb = new StringBuilder(size);

        for(int charCount = 0; charCount < size; charCount++) {
            final int index = random.nextInt(alphaString.length());
            sb.append(alphaString.charAt(index));
        }

        return sb.toString();
    }

    public static String generateEmail() {
        return generateString(10)+"@email.com";
    }

    public static LocalDateTime generateDepartureTime() {
        return LocalDateTime.now()
                .plusDays(random.nextInt(365))
                .withHour(random.nextInt(24))
                .withMinute(random.nextInt(60));
    }

    public static LocalDateTime generateArrivalTime(LocalDateTime departure) {
        return departure
                .plusHours(random.nextInt(15))
                .plusMinutes(random.nextInt(60));
    }

    public static class AirlineDataBuilder {
        private String emailAddress;
        private String departureTime;
        private String airportDepartureCode;
        private String arrivalTime;
        private String airportArrivalCode;
        private String flightNumber;
        private String confirmation;

        public AirlineDataBuilder() {
            LocalDateTime localDepartureTime = generateDepartureTime();
            LocalDateTime localArrivalTime = generateArrivalTime(localDepartureTime);

            this.emailAddress = generateEmail();
            this.departureTime = localDepartureTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
            this.airportDepartureCode = generateAirportCode();
            this.arrivalTime = localArrivalTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
            this.airportArrivalCode = generateAirportCode();
            this.flightNumber = "SKY1" + random.nextInt(1000);
            this.confirmation = "SKY1" + generateString(6);
        }

        public AirlineDataBuilder setEmailAddress(String emailAddress) {
            this.emailAddress = emailAddress;
            return this;
        }

        public AirlineDataBuilder setDepartureTime(String departureTime) {
            this.departureTime = departureTime;
            return this;
        }

        public AirlineDataBuilder setDepartureAirportCode(String airportDepartureCode) {
            this.airportDepartureCode = airportDepartureCode;
            return this;
        }

        public AirlineDataBuilder setArrivalTime(String arrivalTime) {
            this.arrivalTime = arrivalTime;
            return this;
        }

        public AirlineDataBuilder setArrivalAirportCode(String airportArrivalCode) {
            this.airportArrivalCode = airportArrivalCode;
            return this;
        }

        public AirlineDataBuilder setFlightNumber(String flightNumber) {
            this.flightNumber = flightNumber;
            return this;
        }

        public AirlineDataBuilder setConfirmation(String confirmation) {
            this.confirmation = confirmation;
            return this;
        }

        public AirlineData build() {
            AirlineData airlineData = new AirlineData();

            airlineData.setEmailAddress(this.emailAddress);
            airlineData.setDepartureTime(this.departureTime);
            airlineData.setDepartureAirportCode(this.airportDepartureCode);
            airlineData.setArrivalTime(this.arrivalTime);
            airlineData.setArrivalAirportCode(this.airportArrivalCode);
            airlineData.setFlightNumber(this.flightNumber);
            airlineData.setConfirmationCode(this.confirmation);

            return airlineData;
        }
    }

    public static class FlightDataBuilder {
        private String emailAddress;
        private String departureTime;
        private String departureAirportCode;
        private String arrivalTime;
        private String arrivalAirportCode;
        private String flightNumber;
        private String confirmationCode;

        public FlightDataBuilder() {
            LocalDateTime localDepartureTime = generateDepartureTime();
            LocalDateTime localArrivalTime = generateArrivalTime(localDepartureTime);

            this.emailAddress = generateEmail();
            this.departureTime = localDepartureTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
            this.departureAirportCode = generateAirportCode();
            this.arrivalTime = localArrivalTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
            this.arrivalAirportCode = generateAirportCode();
            this.flightNumber = "Flight" + random.nextInt(1000);
            this.confirmationCode = "Confirmation" + generateString(5);
        }

        public FlightDataBuilder setEmailAddress(String emailAddress) {
            this.emailAddress = emailAddress;
            return this;
        }

        public FlightDataBuilder setDepartureTime(LocalDateTime departureTime) {
            this.departureTime = departureTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
            return this;
        }

        public FlightDataBuilder setDepartureAirportCode(String departureAirportCode) {
            this.departureAirportCode = departureAirportCode;
            return this;
        }

        public FlightDataBuilder setArrivalTime(LocalDateTime arrivalTime) {
            this.arrivalTime = arrivalTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
            return this;
        }

        public FlightDataBuilder setArrivalAirportCode(String arrivalAirportCode) {
            this.arrivalAirportCode = arrivalAirportCode;
            return this;
        }

        public FlightDataBuilder setFlightNumber(String flightNumber) {
            this.flightNumber = flightNumber;
            return this;
        }

        public FlightDataBuilder setConfirmationCode(String confirmationCode) {
            this.confirmationCode = confirmationCode;
            return this;
        }

        public FlightData build() {
            FlightData flightData = new FlightData();

            flightData.setEmailAddress(this.emailAddress);
            flightData.setDepartureTime(this.departureTime);
            flightData.setDepartureAirportCode(this.departureAirportCode);
            flightData.setArrivalTime(this.arrivalTime);
            flightData.setArrivalAirportCode(this.arrivalAirportCode);
            flightData.setFlightNumber(this.flightNumber);
            flightData.setConfirmationCode(this.confirmationCode);

            return flightData;
        }
    }

    public static class FlyerStatsDataBuilder {
        private String emailAddress;
        private long totalFlightDuration;
        private long numberOfFlights;

        public FlyerStatsDataBuilder() {
            final LocalDateTime localDepartureTime = generateDepartureTime();
            final LocalDateTime localArrivalTime = generateArrivalTime(localDepartureTime);

            this.emailAddress = generateEmail();
            this.totalFlightDuration = Duration.between(localDepartureTime, localArrivalTime).toMinutes();
            this.numberOfFlights = random.nextInt(5);
        }

        public FlyerStatsDataBuilder setEmailAddress(String emailAddress) {
            this.emailAddress = emailAddress;
            return this;
        }

        public FlyerStatsDataBuilder setTotalFlightDuration(long totalFlightDuration) {
            this.totalFlightDuration = totalFlightDuration;
            return this;
        }

        public FlyerStatsDataBuilder setNumberOfFlights(long numberOfFlights) {
            this.numberOfFlights = numberOfFlights;
            return this;
        }

        public FlyerStatsData build() {
            FlyerStatsData flyerStatsData = new FlyerStatsData();

            flyerStatsData.setEmailAddress(this.emailAddress);
            flyerStatsData.setTotalFlightDuration(this.totalFlightDuration);
            flyerStatsData.setNumberOfFlights(this.numberOfFlights);

            return flyerStatsData;
        }
    }
}

File: python/tests/kickstarter/test_helpers.py
import sys
import os
import random
import datetime
import pytest

# --- Add the source directory containing the 'model' package to the Python path
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '../../src/kickstarter')))

from model.flight_data import FlightData, FlyerStatsData
from model.airline_flight_data import AirlineFlightData


def generate_airport_code():
    airports = [
        "ATL", "DFW", "DEN", "ORD", "LAX", "CLT", "MCO", "LAS", "PHX", "MIA",
        "SEA", "IAH", "JFK", "EWR", "FLL", "MSP", "SFO", "DTW", "BOS", "SLC",
        "PHL", "BWI", "TPA", "SAN", "LGA", "MDW", "BNA", "IAD", "DCA", "AUS"
    ]
    return random.choice(airports)


def generate_string(size):
    alpha_string = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    return ''.join(random.choice(alpha_string) for _ in range(size))


def generate_email():
    return generate_string(10) + "@email.com"


def generate_departure_time():
    now = datetime.datetime.now()
    return now + datetime.timedelta(days=random.randint(0, 364), hours=random.randint(0, 23), minutes=random.randint(0, 59))


def generate_arrival_time(departure_time):
    return departure_time + datetime.timedelta(hours=random.randint(0, 14), minutes=random.randint(0, 59))


# Builder Classes
class AirlineFlightDataBuilder:
    def __init__(self):
        local_departure_time = generate_departure_time()
        local_arrival_time = generate_arrival_time(local_departure_time)

        self.email_address = generate_email()
        self.departure_time = local_departure_time.strftime("%Y-%m-%d %H:%M:%S")
        self.departure_airport_code = generate_airport_code()
        self.arrival_time = local_arrival_time.strftime("%Y-%m-%d %H:%M:%S")
        self.arrival_airport_code = generate_airport_code()
        self.flight_number = "SKY1" + str(random.randint(0, 999))
        self.confirmation_code = "SKY1" + generate_string(6)
        self.ticket_price = random.randint(100, 1000)
        self.aircraft = "Boeing 737"
        self.booking_agency_email = generate_email()

    def set_email_address(self, email_address):
        self.email_address = email_address
        return self

    def set_departure_time(self, departure_time):
        self.departure_time = departure_time
        return self

    def set_departure_airport_code(self, departure_airport_code):
        self.departure_airport_code = departure_airport_code
        return self

    def set_arrival_time(self, arrival_time):
        self.arrival_time = arrival_time
        return self

    def set_arrival_airport_code(self, arrival_airport_code):
        self.arrival_airport_code = arrival_airport_code
        return self

    def set_flight_number(self, flight_number):
        self.flight_number = flight_number
        return self

    def set_confirmation_code(self, confirmation_code):
        self.confirmation_code = confirmation_code
        return self
    
    def set_ticket_price(self, ticket_price):      
        self.ticket_price = ticket_price
        return self
    
    def set_aircraft(self, aircraft):
        self.aircraft = aircraft
        return self
    
    def set_booking_agency_email(self, booking_agency_email):
        self.booking_agency_email = booking_agency_email
        return self

    def build(self):
        return AirlineFlightData(
            email_address=self.email_address,
            departure_time=self.departure_time,
            departure_airport_code=self.departure_airport_code,
            arrival_time=self.arrival_time,
            arrival_airport_code=self.arrival_airport_code,
            flight_number=self.flight_number,
            confirmation_code=self.confirmation_code,
            ticket_price=self.ticket_price,
            aircraft=self.aircraft,
            booking_agency_email=self.booking_agency_email
        )


class FlightDataBuilder:
    def __init__(self):
        local_departure_time = generate_departure_time()
        local_arrival_time = generate_arrival_time(local_departure_time)

        self.email_address = generate_email()
        self.departure_time = local_departure_time.strftime("%Y-%m-%d %H:%M:%S")
        self.departure_airport_code = generate_airport_code()
        self.arrival_time = local_arrival_time.strftime("%Y-%m-%d %H:%M:%S")
        self.arrival_airport_code = generate_airport_code()
        self.flight_number = "Flight" + str(random.randint(0, 999))
        self.confirmation_code = "Confirmation" + generate_string(5)
        self.airline="skyOne"

    def set_email_address(self, email_address):
        self.email_address = email_address
        return self

    def set_departure_time(self, departure_time):
        self.departure_time = departure_time.strftime("%Y-%m-%d %H:%M:%S")
        return self

    def set_departure_airport_code(self, departure_airport_code):
        self.departure_airport_code = departure_airport_code
        return self

    def set_arrival_time(self, arrival_time):
        self.arrival_time = arrival_time.strftime("%Y-%m-%d %H:%M:%S")
        return self

    def set_arrival_airport_code(self, arrival_airport_code):
        self.arrival_airport_code = arrival_airport_code
        return self

    def set_flight_number(self, flight_number):
        self.flight_number = flight_number
        return self

    def set_confirmation_code(self, confirmation_code):
        self.confirmation_code = confirmation_code
        return self
    
    def set_airline(self, airline):
        self.airline = airline
        return self

    def build(self):
        return FlightData(
            email_address=self.email_address,
            departure_time=self.departure_time,
            departure_airport_code=self.departure_airport_code,
            arrival_time=self.arrival_time,
            arrival_airport_code=self.arrival_airport_code,
            flight_number=self.flight_number,
            confirmation_code=self.confirmation_code,
            airline=self.airline
        )


class FlyerStatsDataBuilder:
    def __init__(self):
        local_departure_time = generate_departure_time()
        local_arrival_time = generate_arrival_time(local_departure_time)

        self.email_address = generate_email()
        self.total_flight_duration = int((local_arrival_time - local_departure_time).total_seconds() / 60)
        self.number_of_flights = random.randint(0, 4)

    def set_email_address(self, email_address):
        self.email_address = email_address
        return self

    def set_total_flight_duration(self, total_flight_duration):
        self.total_flight_duration = total_flight_duration
        return self

    def set_number_of_flights(self, number_of_flights):
        self.number_of_flights = number_of_flights
        return self

    def build(self):
        return FlyerStatsData(
            email_address=self.email_address,
            total_flight_duration=self.total_flight_duration,
            number_of_flights=self.number_of_flights
        )


# Pytest test cases
@pytest.fixture
def airline_data_builder():
    return AirlineFlightDataBuilder()


@pytest.fixture
def flight_data_builder():
    return FlightDataBuilder()


@pytest.fixture
def flyer_stats_data_builder():
    return FlyerStatsDataBuilder()


def test_airline_data_builder(airline_data_builder):
    airline_data = airline_data_builder.build()
    assert airline_data.email_address.endswith("@email.com")
    assert len(airline_data.flight_number) > 0
    assert len(airline_data.confirmation_code) > 0


def test_flight_data_builder(flight_data_builder):
    flight_data = flight_data_builder.build()
    assert flight_data.email_address.endswith("@email.com")
    assert len(flight_data.flight_number) > 0
    assert len(flight_data.confirmation_code) > 0


def test_flyer_stats_data_builder(flyer_stats_data_builder):
    flyer_stats_data = flyer_stats_data_builder.build()
    assert flyer_stats_data.total_flight_duration > 0
    assert flyer_stats_data.number_of_flights >= 0
